server/src/main/java/io/druid/client/DirectDruidClient.java

```diff
@@ -306,34 +337,100 @@ public InputStream nextElement()
         @Override
         public void exceptionCaught(final ClientResponse<InputStream> clientResponse, final Throwable e)
         {
-          // Don't wait for lock in case the lock had something to do with the error
-          synchronized (done) {
-            done.set(true);
-            // Make a best effort to put a zero length buffer into the queue in case something is waiting on the take()
-            // If nothing is waiting on take(), this will be closed out anyways.
-            queue.offer(
-                new InputStream()
-                {
-                  @Override
-                  public int read() throws IOException
-                  {
-                    throw new IOException(e);
-                  }
-                }
-            );
+          String msg = StringUtils.safeFormat(
+              "Query[%s] url[%s] failed with exception msg [%s]",
+              query.getId(),
+              url,
+              e.getMessage()
+          );
+          setupResponseReadFailure(msg, e);
+        }
+
+        private void setupResponseReadFailure(String msg, Throwable th)
+        {
+          fail.set(msg);
+          queue.clear();
+          queue.offer(new InputStream()
+          {
+            @Override
+            public int read() throws IOException
+            {
+              if (th != null) {
+                throw new IOException(msg, th);
+              } else {
+                throw new IOException(msg);
+              }
+            }
+          });
+
+        }
+
+        // Returns remaining timeout or throws exception if timeout already elapsed.
+        private long checkQueryTimeout()
+        {
+          Object obj = context.get(QUERY_START_TIME);
+
+          if (obj != null) {
+            long timeLeft = QueryContexts.getTimeout(query) - (System.currentTimeMillis() - ((Long) obj).longValue());
+            if (timeLeft <= 0) {
+              String msg = StringUtils.safeFormat("Query[%s] url[%s] timed out.", query.getId(), url);
+              setupResponseReadFailure(msg, null);
+              throw new RE(msg);
+            } else {
+              return timeLeft;
+            }
+          } else {
+            return QueryContexts.getTimeout(query);
+          }
+        }
+
+        private void checkTotalBytesLimit(long bytes)
+        {
+          long limit = QueryContexts.getMaxScatterGatherBytes(query);
+
+          if (limit < Long.MAX_VALUE) {
+            synchronized (context) {
+              if(context.get(QUERY_TOTAL_BYTES_GATHERED) == null) {
```

Comment:
I think it would be better to create this once higher up rather than checked and double checked constantly with everything.